### PR TITLE
Bug fix for Workload controller Manager not stable issue

### PR DESCRIPTION
**Pull request by Sindica**
Issue https://github.com/futurewei-cloud/arktos/issues/523

There are 4 commits.
1st fix health check issue in perf test
2nd fix health report issue in Arktos
3rd simplified the workload distribution logic to always evenly distribute the workload based on controller instance id ordering. This algorithm will be revisit and optimize later for traffic reducing.
4th use dedicated rest client for controller health report - avoid throttling with shared client

### CODE REVIEW

**Comment by Sindica** [comment]
Test done:
1. Checked RateLimiter pointer
2. Perf test density for 500 node, duration 1h40m. Controller instances have no changes.

File: pkg/cloudfabric-controller/controllerframework/controller_framework.go [hunk 1]
```diff
@@ -199,24 +169,28 @@ func (c *ControllerBase) WatchInstanceUpdate(stopCh <-chan struct{}) {
...
+	klog.Infof("Got controller instance update massage. Updated Controller Type %s, current controller instance type %s, key %d", updatedType, c.controllerType, c.controllerKey)
+	if updatedType != c.controllerType {
+		return
+	}
+
+	klog.V(4).Infof("Start updating controller instance %s", c.controllerType)
+	controllerInstances, err := listControllerInstancesByType(c.controllerType)
+	if err != nil {
+		// TODO - add retry
+		klog.Errorf("Error getting controller instances for %s. Error %v", c.controllerType, err)
```
Comment (chenqianfzh) [inline pkg/cloudfabric-controller/controllerframework/controller_framework.go:187]: Nit: the log message should be " Error getting controller instances for type %s". It looks like %s is the instance name now.

File: pkg/cloudfabric-controller/controllerframework/controller_framework.go [hunk 2]
```diff
@@ -199,24 +169,28 @@ func (c *ControllerBase) WatchInstanceUpdate(stopCh <-chan struct{}) {
...
+		return
+	}
+	c.updateCachedControllerInstances(controllerInstances)
+	klog.V(4).Infof("Done updating controller instance %s", c.controllerType)
```
Comment (chenqianfzh) [inline pkg/cloudfabric-controller/controllerframework/controller_framework.go:191]: Nit: log:  "Done updating controller of type %s"

File: pkg/cloudfabric-controller/controllerframework/controller_framework.go [hunk 3]
```diff
@@ -338,57 +265,31 @@ func (c *ControllerBase) updateCachedControllerInstances(newControllerInstanceMa
...
-			klog.Infof("Self range updated %d, %d", newLowerBound, newUpperbound)
 		}
 
 		message := fmt.Sprintf("Controller %s instance %s", c.controllerType, c.controllerName)
 		if currentControllerInstance.lowerboundKey != newLowerBound || currentControllerInstance.controllerKey != newUpperbound {
-			if c.sortedControllerInstancesLocal[c.curPos].lowerboundKey == 0 {
-				message += fmt.Sprintf(" old range [%d, %d]", c.sortedControllerInstancesLocal[c.curPos].lowerboundKey, c.sortedControllerInstancesLocal[c.curPos].controllerKey)
-			} else {
-				message += fmt.Sprintf(" old range (%d, %d]", c.sortedControllerInstancesLocal[c.curPos].lowerboundKey, c.sortedControllerInstancesLocal[c.curPos].controllerKey)
+			if c.curPos >= 0 {
```
Comment (chenqianfzh) [inline pkg/cloudfabric-controller/controllerframework/controller_framework.go:287]: Shall the code just return if c.curPos < 0, as it means the registration is incomplete?

File: pkg/cloudfabric-controller/controllerframework/controller_framework_utils.go [hunk 4]
```diff
@@ -54,6 +53,31 @@ func SortControllerInstancesByKeyAndConvertToLocal(controllerInstanceMap map[str
 	return sortedControllerInstancesLocal
 }
 
+// evenly split 0 ~ maxInt64 interval
+func reassignControllerKeys(instances []controllerInstanceLocal) []controllerInstanceLocal {
+	if len(instances) == 0 {
+		return instances
+	}
+
+	if len(instances) == 1 {
```
Comment (chenqianfzh) [inline pkg/cloudfabric-controller/controllerframework/controller_framework_utils.go:62]: line 62-64 can be deleted. the logic in line 67-76 will handle the case that len(instances) ==0.

File: pkg/cloudfabric-controller/controllerframework/controller_framework.go [hunk 5]
```diff
@@ -632,9 +440,18 @@ func (c *ControllerBase) ReportHealth() {
...
 }
 
-func generateControllerName() string {
-	uid := uuid.NewUUID()
-	return fmt.Sprintf("%s-%v", controllerInstanceNamePrefix, uid)
+func generateControllerName(existedInstanceMap map[string]v1.ControllerInstance) string {
+	for {
+		uid := uuid.NewUUID()
+		name := fmt.Sprintf("%s-%v", controllerInstanceNamePrefix, uid)
+		_, isExist := existedInstanceMap[name]
```
Comment (chenqianfzh) [inline pkg/cloudfabric-controller/controllerframework/controller_framework.go:448]: Retry should solve the problem. The chance that a unique name can not be created after a few retrials is almost zero.

File: pkg/cloudfabric-controller/controllerframework/controller_framework.go [hunk 6]
```diff
@@ -403,94 +304,51 @@ func (c *ControllerBase) updateCachedControllerInstances(newControllerInstanceMa
...
-	}
+// 		isUpdate: 					controller instances were changed
+//		isSelfUpdate:   			is current controller instance updated (include lowerbound, upperbound, islocked)
+//		newLowerBound:  			new lowerbound for current controller instance
+//		newUpperBound:  			new upperbound for current controller instance
+//		newPos:						new position of current controller instance in sorted controller list (sorted by controllerKey - upperbound)
+//		updatedControllerInstances: new sorted controller instances
+//
+// Note 08-06-2020: Simplify the scenario to always evenly redistribute the entire problem space first
+// Will need to revisit and design for long term solution later after release 0.3 (2020/0730 release)
```
Comment (chenqianfzh) [inline pkg/cloudfabric-controller/controllerframework/controller_framework.go:347]: It is better to create a tracking issue.

Comment (Sindica) [reply pkg/cloudfabric-controller/controllerframework/controller_framework.go:347]: > It is better to create a tracking issue.

There is one: https://github.com/futurewei-cloud/arktos/issues/63

File: pkg/cloudfabric-controller/controllerframework/controller_framework.go (hunk 5 above)
Comment (Sindica) [reply pkg/cloudfabric-controller/controllerframework/controller_framework.go:448]: > Retry should solve the problem. The chance that a unique name can not be created after a few retrials is almost zero.

This is retry. Right?

File: pkg/cloudfabric-controller/controllerframework/controller_framework_utils.go (hunk 4 above)
Comment (Sindica) [reply pkg/cloudfabric-controller/controllerframework/controller_framework_utils.go:62]: Yes. And there is test cover this scenario. Removed.

File: pkg/cloudfabric-controller/controllerframework/controller_framework.go (hunk 3 above)
Comment (Sindica) [reply pkg/cloudfabric-controller/controllerframework/controller_framework.go:287]: > Shall the code just return if c.curPos < 0, as it means the registration is incomplete?

Yes, this code is also used in registration process. It needs to handle curPos < 0

File: pkg/cloudfabric-controller/controllerframework/controller_framework.go (hunk 1 above)
Comment (Sindica) [reply pkg/cloudfabric-controller/controllerframework/controller_framework.go:187]: instances is a plural. It means type already. I updated ling 183 to "Start updating controller type %s"

**Pull request merged**